test_suite/tests/derive.rs

```diff
@@ -179,3 +179,120 @@ fn fields_with_type_alias() {
 
     assert_type!(S, ty);
 }
+
+#[test]
+fn associated_types_derive_without_bounds() {
+    trait Types {
+        type A;
+    }
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct Assoc<T: Types> {
+        a: T::A,
+    }
+
+    #[derive(TypeInfo)]
+    enum ConcreteTypes {}
+    impl Types for ConcreteTypes {
+        type A = bool;
+    }
+
+    let struct_type = Type::builder()
+        .path(Path::new("Assoc", "derive"))
+        .type_params(tuple_meta_type!(ConcreteTypes))
+        .composite(Fields::named().field_of::<bool>("a", "T::A"));
+
+    assert_type!(Assoc<ConcreteTypes>, struct_type);
+}
+
+#[test]
+fn adds_proper_trait_bounds() {
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct Cat {
+        tail: bool,
+        ears: u8,
+    }
+    fn assert_type_info<T: TypeInfo + 'static>() {};
+
+    assert_type_info::<Cat>();
+}
+
+#[test]
+fn adds_type_info_trait_bounds_for_all_generics() {
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    enum PawType<Paw> {
+        Big(Paw),
+        Small(Paw),
+    }
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct Cat<Tail, Ear, Paw> {
+        tail: Tail,
+        ears: [Ear; 3],
+        paws: PawType<Paw>,
+    }
+
+    fn assert_type_info<T: TypeInfo + 'static>() {};
+    assert_type_info::<Cat<bool, u8, u16>>();
+}
+
+#[test]
+fn adds_correct_bounds_to_self_referential_types() {
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct Nested<P> {
+        pos: P,
+    };
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct Is<N> {
+        nexted: N,
+    };
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct That<I, S> {
+        is: I,
+        selfie: S,
+    };
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct Thing<T> {
+        that: T,
+    };
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct Other<T> {
+        thing: T,
+    };
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct Selfie<Pos> {
+        another: Box<Selfie<Pos>>,
+        pos: Pos,
+        nested: Box<Other<Thing<That<Is<Nested<Pos>>, Selfie<Pos>>>>>,
+    }
+
+    fn assert_type_info<T: TypeInfo + 'static>() {};
+    assert_type_info::<Selfie<bool>>();
+}
+#[test]
+fn self_referential() {
+    #[allow(unused)]
+    #[derive(TypeInfo)]
+    struct Meee {
+        me: Box<Meee>,
+    }
+    fn assert_type_info<T: TypeInfo + 'static>() {};
+    assert_type_info::<Meee>();
+}
+
+#[rustversion::nightly]
```

Comment:
Error messages are quite different on stable currently and as we all (?) use nightly for development I went for this.